### PR TITLE
chore(shell-api): allow more advanced interrupt handling MONGOSH-1555

### PR DESCRIPTION
**Pull request by addaleax**

### CODE REVIEW

File: packages/shell-api/src/interruptor.ts [hunk 1]
```diff
@@ -75,12 +86,16 @@ export class InterruptFlag {
...
+   * installed through withOverrideInterruptBehavior() may perform additional
+   * cleanup work before the current connection is ready to be severed.
    */
-  public set(): void {
+  public async set(): Promise<void> {
     this.interrupted = true;
     const err = new MongoshInterruptedError();
     for (const listener of [...this.onInterruptListeners]) {
-      listener(err);
+      await listener(err);
```
Comment (shaketbaby) [inline packages/shell-api/src/interruptor.ts:98]: Do we want to catch any errors that the listener might throw so all listeners will always be notified?
```ts
try {
  await listener(err);
} catch {
  // ignore
}
```

**Review by shaketbaby — APPROVED** [review]
Looks good, thanks for doing this 👍

File: packages/shell-api/src/interruptor.ts (hunk 1 above)
Comment (addaleax) [reply packages/shell-api/src/interruptor.ts]: Yes, that seems like a good idea :+1:

**Comment by addaleax** [comment]
Evergreen: https://spruce.mongodb.com/version/64ec68b561837d838d2a6308/tasks

**Pull request merged**